consensus/state.go

```diff
@@ -1363,11 +1359,67 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
+		24: broadcast <PREVOTE, h_p, round_p, id(v)>
+
+		Here, cs.Proposal.POLRound corresponds to the -1 in the above algorithm rule.
+		This means that the proposer is producing a new proposal that has not previously
+		seen a 2/3 majority by the network.
+
+		If we have already locked on a different value that is different from the proposed value,
+		we prevote nil since we are locked on a different value. Otherwise, if we're not locked on a block
+		or the proposal matches our locked block, we prevote the proposal.
+	*/
+	if cs.Proposal.POLRound == -1 {
+		if cs.LockedRound == -1 {
+			logger.Debug("prevote step: ProposalBlock is valid and there is no locked block; prevoting the proposal")
+			cs.signAddVote(cmtproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+			return
+		}
+		if cs.ProposalBlock.HashesTo(cs.LockedBlock.Hash()) {
+			logger.Debug("prevote step: ProposalBlock is valid and matches our locked block; prevoting the proposal")
+			cs.signAddVote(cmtproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+			return
+		}
+	}
+
+	/*
+		28: upon <PROPOSAL, h_p, round_p, v, v_r> from proposer(h_p, round_p) AND 2f + 1 <PREVOTE, h_p, v_r, id(v)> while
+		step_p = propose && (v_r ≥ 0 && v_r < round_p) do
+		29: if valid(v) && (lockedRound_p ≤ v_r || lockedValue_p = v) then
+		30: broadcast <PREVOTE, h_p, round_p, id(v)>
+
+		This rule is a bit confusing but breaks down as follows:
+
+		If we see a proposal in the current round for value 'v' that lists its valid round as 'v_r'
+		AND this validator saw a 2/3 majority of the voting power prevote 'v' in round 'v_r', then we will
+		issue a prevote for 'v' in this round if 'v' is valid and either matches our locked value OR
+		'v_r' is a round greater than or equal to our current locked round.
+
+		'v_r' can be a round greater than to our current locked round if a 2/3 majority of
+		the network prevoted a value in round 'v_r' but we did not lock on it, possibly because we
+		missed the proposal in round 'v_r'.
+	*/
+	blockID, ok := cs.Votes.Prevotes(cs.Proposal.POLRound).TwoThirdsMajority()
+	if ok && cs.ProposalBlock.HashesTo(blockID.Hash) && cs.Proposal.POLRound >= 0 && cs.Proposal.POLRound < cs.Round {
+		if cs.LockedRound <= cs.Proposal.POLRound {
+			logger.Debug("prevote step: ProposalBlock is valid and received a 2/3" +
+				"majority in a round later than the locked round; prevoting the proposal")
+			cs.signAddVote(cmtproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+			return
+		}
+		if cs.ProposalBlock.HashesTo(cs.LockedBlock.Hash()) {
+			logger.Debug("prevote step: ProposalBlock is valid and matches our locked block; prevoting the proposal")
+			cs.signAddVote(cmtproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+			return
+		}
+	}
+
+	logger.Debug("prevote step: ProposalBlock is valid but was not our locked block or" +
```

Comment:
What should happen at this point when `!ok` or `blockID.IsNil()`? According to the paper, the rule at line 28 should not apply, so here we should probably _not_ vote anything (and log something).

Reply:
My understanding of function `defaultDoPrevote` is that it will be called when we are ready to send our prevote:

* either because we detected a majority (outside `defaultDoPrevote`),
* or because we timed out

hence, we will prevote nil, rather than keep waiting.

Although I'm not 100% sure of this, I should probably examine all call sites.

Reply:
I have the same understanding.
If the timeout hit before having the full proposal, prevote nil at the beginning.
If the timeout hit before having a majority, prevote nil here.

Btw, in [state.go](https://github.com/cometbft/cometbft/blob/48de47782c2fb06d620b6825c5dc1e074a724aff/consensus/state.go#L1285-L1286), update the comment // Prevote for ~LockedBlock if we're locked, or ~ProposalBlock if valid.

Reply:
> Btw, in [state.go](https://github.com/cometbft/cometbft/blob/48de47782c2fb06d620b6825c5dc1e074a724aff/consensus/state.go#L1285-L1286), update the comment // Prevote for ~LockedBlock if we're locked, or ~ProposalBlock if valid.

But we're doing it already, aren't we?

Reply:
This part of the comment is no longer true `Prevote for LockedBlock if we're locked`; this was the first if that you removed.

Reply:
Pls go ahead with the merge